lib/ui/painting.dart

```diff
@@ -1215,6 +1225,25 @@ class Image extends NativeFieldWrapperClass2 {
   /// The number of image pixels along the image's vertical axis.
   int get height native 'Image_height';
 
+  /// Converts the [Image] object into a byte array.
+  ///
+  /// The [format] is encoding format to be used.
+  ///
+  /// The [quality] is a value in the range 0 to 100. [quality] is a format specific metric trading
+  /// off size and encoding error. When used, 100 encodes with highest quality (i.e., least error)
+  /// and 0 encodes with lowest quality (i.e., smallest size).
+  ///
+  /// Returns a future which complete with the binary image data (e.g a PNG or JPEG binary data) or
+  /// an error if encoding fails.
+  Future<Uint8List> toByteData({EncodeFormat format: EncodeFormat.JPEG, int quality: 80}) {
```

Comment:
It feels a little weird to pass quality when it doesn't always make sense depending on the format.  What if the API were as follows:

```dart
/// The encoding formats supported by the [Image.toByteData].
class EncodingFormat {
  const EncodingFormat._(this._format, [this._quality = 0]);

  final int _format;
  final int _quality;

  // Be conservative with the formats we expose. It is easy to add new formats in future but
  // difficult to remove.
  // These values must be kept in sync with the logic in ToSkEncodedImageFormat.
  static const int _jpegFormat = 0;
  static const int _pngFormat = 1;
  static const int _webpFormat = 2;

  /// PNG format.
  ///
  /// A loss-less compression format for images. This format is well suited for images
  /// with hard edges, such as screenshots or sprites, and images with text. Transparency
  /// is supported. The `quality` metric is ignored for this format. The PNG format supports
  /// images up to 2,147,483,647 pixels in either dimension, though in practice available
  /// memory provides a more immediate limitation on maximum image size.
  ///
  /// PNG images normally use the `.png` file extension and the `image/png` MIME type.
  ///
  /// See also:
  ///
  ///  * <https://en.wikipedia.org/wiki/Portable_Network_Graphics>, the Wikipedia page on PNG.
  ///  * <https://tools.ietf.org/rfc/rfc2083.txt>, the PNG standard.
  static const EncodingFormat png = const EncodingFormat._(_pngFormat);

  /// JPEG format.
  ///
  /// This format, strictly speaking called JFIF, is a lossy compression graphics format that
  /// can handle images up to 65,535 pixels in either dimension. The [quality] metric controls
  /// the compression ratio. Values in the range of about 50 to 90 are somewhat reasonable;
  /// values above 95 increase the file size with little noticeable improvement to the quality,
  /// values below 50 drop the quality substantially.
  ///
  /// This format is well suited for photographs. It is very poorly suited for images with hard
  /// edges or text. It does not support transparency.
  ///
  /// JPEG images normally use the `.jpeg` file extension and the `image/jpeg` MIME type.
  ///
  /// See also:
  ///
  ///  * <https://en.wikipedia.org/wiki/JPEG>, the Wikipedia page on JPEG.
  factory EncodingFormat.jpeg([int quality = 80]) => new EncodingFormat._(_jpegFormat, quality);

  /// WebP format.
  ///
  /// The WebP format supports both lossy and lossless compression; however, the
  /// [Image.toByteData] method always uses lossy compression when [webp] is
  /// specified. The [quality] metric determines the compression ratio; higher values
  /// result in better quality but larger file sizes, and vice versa. WebP images are
  /// limited to 16,383 pixels in each direction (width and height).
  ///
  /// WebP images normally use the `.webp` file extension and the `image/webp` MIME type.
  ///
  /// See also:
  ///
  ///  * <https://en.wikipedia.org/wiki/WebP>, the Wikipedia page on WebP.
  factory EncodingFormat.webp([int quality = 80]) => new EncodingFormat._(_webpFormat, quality);
}

Future<Uint8List> toByteData({EncodingFormat format: EncodingFormat.png}) {
  return _futurize(
        (_Callback<Uint8List> callback) => _toByteData(format._format, format._quality, callback)
  );
```

Reply:
We could even then provide some suggested values...

```dart
  static const EncodingFormat jpegHigh = const EncodingFormat._(_jpegFormat, 95);
  static const EncodingFormat jpegMedium = const EncodingFormat._(_jpegFormat, 80);
  static const EncodingFormat jpegLow = const EncodingFormat._(_jpegFormat, 60);
```

Reply:
I like this quite a bit more. It provides a much nicer API and enables to have different default quality values for different formats. 

nit: I will probably default the quality to be '100' when unspecified. It is safer if Skia ever decides to respect quality for a format in future.

nit: If we decide to provide additional (Low,Medium, High) suggested values we probably should
do it consistently for all formats where quality is supported.

Reply:
👍

Reply:
I like this proposal, though please use named arguments not optional arguments for the quality (easier to extend, easier to understand -- a number alone could be any number of things).

We may want to consider making the PNG case a constructor too, so that we can add arguments later.

We don't have to use factory constructors, we could just use real constructors. Then we don't have to have the private constructor.

I think if we're going to have a fallback default, 0 is safer, because it's more likely to cause noticeable problems (like crashes, divide by zero, horrible output, etc), whereas 100 is more likely to cause very subtle problems (like very large file sizes).

Reply:
I probably wouldn't bother with the low/medium/high defaults. I'm not sure what problem they really solve.

Reply:
Done.  Used a class with three regular constructors, one for each supported format.